File: compilers/go/codegen/r119_field_canonicity_test.go
package codegen

import (
	"math/big"
	"testing"
)

// ---------------------------------------------------------------------------
// R-119 -- no canonicity or range check on any witness-supplied field element.
//
// WHAT REPRODUCED, AND WHAT DID NOT. The finding as filed names two shapes,
// `v >= p` and `v < 0`, and asks that `v + p` either be rejected or "produce
// the same verdict as v at every equality site". The SECOND disjunct already
// held for `v + p`: every emitter here reduces its result mod p, so measured on
// the go-sdk interpreter before this gate,
//
//	bbFieldAdd(5+p, 0) -> 5      bbFieldAdd(5, 0) -> 5
//	bbFieldSub(5+p, 0) -> 5      bbFieldInv(3+p)  -> 1342177281 = bbFieldInv(3)
//
// so `v + p` was already indistinguishable from `v`. It is gated here anyway,
// because leaving two accepted spellings of one element is the aliasing the
// finding is about and because the gate is one OP_WITHIN either way -- but the
// half that was actually BROKEN is the negative one:
//
//	bbFieldAdd(-1, 0)  -> -1            bbFieldAdd(p-1, 0)  -> 2013265920
//	bbFieldMul(-1, 1)  -> -1            bbFieldMul(p-1, 1)  -> 2013265920
//	bbFieldInv(-1)     -> -1            bbFieldInv(p-1)     -> 2013265920
//	kbExt4Mul0((0,-1,0,0),(0,0,0,1)) -> -3    canonical spelling -> 2130706430
//
// `bbFieldAdd` and `bbFieldMul` reduce with a BARE OP_MOD, on the documented
// assumption that both operands are already in [0, p-1] ("Sum of two values in
// [0, p-1] is always non-negative, so simple OP_MOD suffices"). OP_MOD takes the
// sign of the DIVIDEND, so a negative operand walks straight out of the field:
// the builtin returns a number that is congruent to the right answer and is NOT
// the right answer's script encoding. Script equality is numeric, so the escaped
// spelling breaks every downstream `===`, every OP_NUM2BIN of a field element,
// and the closure property the ext4 and Poseidon2 code depends on when it feeds
// one field builtin into the next.
//
// The gate, per witness operand, on the public entry points only:
//
//	OP_DUP/OP_PICK <0> <p> OP_WITHIN OP_VERIFY
//
// The two facts below are pinned as tests, not as prose: that `v + p` was
// ALREADY equivalent to `v` (so the fix is a tightening, not a correctness
// repair, on that half) and that a negative operand was NOT.
// ---------------------------------------------------------------------------

type fieldTier struct {
	name string
	p    *big.Int
	add  func(func(StackOp))
	sub  func(func(StackOp))
	mul  func(func(StackOp))
	inv  func(func(StackOp))
	// ext4Mul takes 8 operands, ext4Inv takes 4.
	ext4Mul func(func(StackOp))
	ext4Inv func(func(StackOp))
}

func fieldTiers() []fieldTier {
	return []fieldTier{
		{"babybear", bbFieldP, EmitBBFieldAdd, EmitBBFieldSub, EmitBBFieldMul, EmitBBFieldInv,
			EmitBBExt4Mul0, EmitBBExt4Inv0},
		{"koalabear", kbFieldP, EmitKBFieldAdd, EmitKBFieldSub, EmitKBFieldMul, EmitKBFieldInv,
			EmitKBExt4Mul0, EmitKBExt4Inv0},
	}
}

// runField pushes `in`, runs `emitFn`, compares the single result against
// `want` with OP_NUMEQUAL, and reports whether the script ACCEPTED. Every
// classification in this file is the interpreter's error, never prose.
func runField(t *testing.T, emitFn func(func(StackOp)), in []*big.Int, want *big.Int) bool {
	t.Helper()
	var ops []StackOp
	for _, v := range in {
		ops = append(ops, pushBigInt(v))
	}
	ops = append(ops, gatherOps(emitFn)...)
	ops = append(ops, pushBigInt(want), opcode("OP_NUMEQUAL"))
	return BuildAndExecuteOps(ops) == nil
}

// runFieldRaw runs the emitter and asserts nothing about the result — used
// where only accept/reject matters.
func runFieldRaw(t *testing.T, emitFn func(func(StackOp)), in []*big.Int) bool {
	t.Helper()
	var ops []StackOp
	for _, v := range in {
		ops = append(ops, pushBigInt(v))
	}
	ops = append(ops, gatherOps(emitFn)...)
	ops = append(ops, opcode("OP_DROP"), opcode("OP_1"))
	return BuildAndExecuteOps(ops) == nil
}

func bi(n int64) *big.Int { return big.NewInt(n) }

// TestR119_NonCanonicalScalarOperandRejected -- the finding, on the four
// scalar builtins of both fields.
func TestR119_NonCanonicalScalarOperandRejected(t *testing.T) {
	for _, F := range fieldTiers() {
		t.Run(F.name, func(t *testing.T) {
			p := F.p
			pm1 := new(big.Int).Sub(p, bi(1))
			aliasA := new(big.Int).Add(bi(5), p) // 5 + p
			big2p := new(big.Int).Add(p, p)      // 2p, i.e. 0 + 2p

			type binCase struct {
				name string
				emit func(func(StackOp))
			}
			for _, c := range []binCase{
				{"add", F.add}, {"sub", F.sub}, {"mul", F.mul},
			} {
				// >= p on either operand.
				if runFieldRaw(t, c.emit, []*big.Int{aliasA, bi(1)}) {
					t.Errorf("%s(5+p, 1) accepted", c.name)
				}
				if runFieldRaw(t, c.emit, []*big.Int{bi(1), aliasA}) {
					t.Errorf("%s(1, 5+p) accepted", c.name)
				}
				if runFieldRaw(t, c.emit, []*big.Int{big2p, bi(1)}) {
					t.Errorf("%s(2p, 1) accepted", c.name)
				}
				// negative on either operand -- the half that returned a
				// non-canonical result rather than merely an alias.
				if runFieldRaw(t, c.emit, []*big.Int{bi(-1), bi(0)}) {
					t.Errorf("%s(-1, 0) accepted", c.name)
				}
				if runFieldRaw(t, c.emit, []*big.Int{bi(0), bi(-1)}) {
					t.Errorf("%s(0, -1) accepted", c.name)
				}
				// exactly p is the first rejected value; p-1 is the last legal one.
				if runFieldRaw(t, c.emit, []*big.Int{p, bi(0)}) {
					t.Errorf("%s(p, 0) accepted", c.name)
				}
			}

			if runFieldRaw(t, F.inv, []*big.Int{aliasA}) {
				t.Errorf("inv(5+p) accepted")
			}
			if runFieldRaw(t, F.inv, []*big.Int{bi(-1)}) {
				t.Errorf("inv(-1) accepted")
			}
			if runFieldRaw(t, F.inv, []*big.Int{p}) {
				t.Errorf("inv(p) accepted")
			}
			// CONTROL: p-1 is a legal field element on every entry point.
			if !runFieldRaw(t, F.inv, []*big.Int{pm1}) {
				t.Errorf("CONTROL inv(p-1) rejected")
			}
		})
	}
}

// TestR119_ScalarControlsStillVerify -- teeth. An over-strict gate reddens
// here, and every accept is checked against a value computed OFF-CHAIN.
func TestR119_ScalarControlsStillVerify(t *testing.T) {
	for _, F := range fieldTiers() {
		t.Run(F.name, func(t *testing.T) {
			p := F.p
			pm1 := new(big.Int).Sub(p, bi(1))
			mod := func(x *big.Int) *big.Int { return new(big.Int).Mod(x, p) }

			for _, ab := range [][2]*big.Int{
				{bi(0), bi(0)}, {bi(0), pm1}, {pm1, bi(0)}, {pm1, pm1},
				{bi(5), bi(7)}, {bi(1), pm1},
			} {
				a, b := ab[0], ab[1]
				if !runField(t, F.add, []*big.Int{a, b}, mod(new(big.Int).Add(a, b))) {
					t.Errorf("CONTROL add(%s, %s) wrong or rejected", a, b)
				}
				if !runField(t, F.sub, []*big.Int{a, b}, mod(new(big.Int).Sub(a, b))) {
					t.Errorf("CONTROL sub(%s, %s) wrong or rejected", a, b)
				}
				if !runField(t, F.mul, []*big.Int{a, b}, mod(new(big.Int).Mul(a, b))) {
					t.Errorf("CONTROL mul(%s, %s) wrong or rejected", a, b)
				}
			}
			// inv: a * inv(a) == 1 for a legal non-zero a, computed off-chain.
			for _, a := range []*big.Int{bi(1), bi(3), bi(12345), pm1} {
				want := new(big.Int).Exp(a, new(big.Int).Sub(p, bi(2)), p)
				if !runField(t, F.inv, []*big.Int{a}, want) {
					t.Errorf("CONTROL inv(%s) wrong or rejected", a)
				}
			}
			// inv(0) = 0 stays defined, as Fermat gives.
			if !runField(t, F.inv, []*big.Int{bi(0)}, bi(0)) {
				t.Errorf("CONTROL inv(0) wrong or rejected")
			}
		})
	}
}

// TestR119_Ext4OperandsGated -- the eight- and four-operand entry points.
func TestR119_Ext4OperandsGated(t *testing.T) {
	for _, F := range fieldTiers() {
		t.Run(F.name, func(t *testing.T) {
			p := F.p
			pm1 := new(big.Int).Sub(p, bi(1))
			ok8 := []*big.Int{bi(1), bi(2), bi(3), bi(4), bi(5), bi(6), bi(7), pm1}
			ok4 := []*big.Int{bi(1), bi(2), bi(3), pm1}

			// CONTROL: a fully canonical operand vector still runs.
			if !runFieldRaw(t, F.ext4Mul, ok8) {
				t.Fatalf("CONTROL ext4Mul0 on canonical operands rejected")
			}
			if !runFieldRaw(t, F.ext4Inv, ok4) {
				t.Fatalf("CONTROL ext4Inv0 on canonical operands rejected")
			}

			// Each of the eight positions must be gated independently — a gate
			// on "the first operand" only would pass a weaker test.
			for i := range ok8 {
				for _, bad := range []*big.Int{new(big.Int).Add(ok8[i], p), bi(-1), p} {
					in := append([]*big.Int{}, ok8...)
					in[i] = bad
					if runFieldRaw(t, F.ext4Mul, in) {
						t.Errorf("ext4Mul0 accepted %s at operand %d", bad, i)
					}
				}
			}
			for i := range ok4 {
				for _, bad := range []*big.Int{new(big.Int).Add(ok4[i], p), bi(-1), p} {
					in := append([]*big.Int{}, ok4...)
					in[i] = bad
					if runFieldRaw(t, F.ext4Inv, in) {
						t.Errorf("ext4Inv0 accepted %s at operand %d", bad, i)
					}
				}
			}
		})
	}
}

// TestR119_GateIsIdempotentUnderComposition -- the reason the gate goes on the
// INPUT rather than fixing up the output: a gated builtin returns a canonical
// element, so feeding one into the next passes the next gate for free. If this
// reddens, the output escaped the field somewhere.
func TestR119_GateIsIdempotentUnderComposition(t *testing.T) {
	for _, F := range fieldTiers() {
		t.Run(F.name, func(t *testing.T) {
			p := F.p
			pm1 := new(big.Int).Sub(p, bi(1))
			// sub(add(a, b), b) == a, executed as one script: the inner result
			// is fed straight into the outer builtin's gate.
			for _, ab := range [][2]*big.Int{{pm1, pm1}, {bi(0), pm1}, {bi(5), bi(7)}} {
				a, b := ab[0], ab[1]
				var ops []StackOp
				ops = append(ops, pushBigInt(a), pushBigInt(b))
				ops = append(ops, gatherOps(F.add)...) // [add]
				ops = append(ops, pushBigInt(b))
				ops = append(ops, gatherOps(F.sub)...) // [sub(add, b)]
				ops = append(ops, pushBigInt(a), opcode("OP_NUMEQUAL"))
				if err := BuildAndExecuteOps(ops); err != nil {
					t.Errorf("sub(add(%s, %s), %s) != %s: %v", a, b, b, a, err)
				}
			}
		})
	}
}

File: packages/runar-testing/src/__tests__/r119-field-canonicity.test.ts
import { describe, it, expect } from 'vitest';
import {
  emitMethod,
  emitBBFieldAdd, emitBBFieldSub, emitBBFieldMul, emitBBFieldInv,
  emitBBExt4Mul0, emitBBExt4Inv0,
  emitKBFieldAdd, emitKBFieldSub, emitKBFieldMul, emitKBFieldInv,
  emitKBExt4Mul0, emitKBExt4Inv0,
} from 'runar-compiler';
import type { StackOp } from 'runar-ir-schema';
import { ScriptVM, TestContract } from '../index.js';

/**
 * R-119 — no canonicity or range check on ANY witness-supplied field element.
 *
 * `babybear-codegen.ts` and `koalabear-codegen.ts` (and their four ports)
 * emitted only OP_ADD / OP_SUB / OP_MUL / OP_MOD plus shifts and alt-stack
 * traffic. The OP_LESSTHAN / OP_WITHIN / OP_GREATERTHANOREQUAL count in both
 * files was ZERO. Every operand of the four scalar builtins and the eight ext4
 * entry points is an unlock argument, so a witness element >= p, or negative,
 * flowed into the arithmetic unchallenged.
 *
 * WHAT REPRODUCED, AND WHAT DID NOT — the finding names two shapes, and they
 * are not equally broken.
 *
 *   `v + p`: NOT broken. Every emitter reduces its result mod p, so measured on
 *   @bsv/sdk's Spend before this gate, `bbFieldAdd(5+p, 0)` and
 *   `bbFieldAdd(5, 0)` both returned 5, and `bbFieldInv(3+p)` equalled
 *   `bbFieldInv(3)`. On the finding's own disjunction ("rejected, or the same
 *   verdict as v at every equality site") the second disjunct already held.
 *
 *   NEGATIVE: broken. `fieldAdd` and `fieldMul` reduce with a BARE OP_MOD, on
 *   the documented assumption that both operands are already in [0, p-1]
 *   ("Sum of two values in [0, p-1] is always non-negative, so simple OP_MOD
 *   suffices"). OP_MOD takes the sign of the DIVIDEND. Measured:
 *
 *       bbFieldAdd(-1, 0) -> -1          bbFieldAdd(p-1, 0) -> 2013265920
 *       bbFieldMul(-1, 1) -> -1          bbFieldInv(-1)     -> -1
 *       kbExt4Mul0((0,-1,0,0),(0,0,0,1)) -> -3   canonical -> 2130706430
 *
 *   The builtin returned a number congruent to the right answer that is NOT the
 *   right answer's script encoding. Script equality is numeric, so the escaped
 *   spelling breaks every downstream `===`, every OP_NUM2BIN of the element,
 *   and the closure property the ext4 / Poseidon2 code relies on when it feeds
 *   one field builtin into the next.
 *
 * THE GATE REJECTS rather than reduces, and sits on the INPUT: a reduce would
 * leave `v` and `v + p` as two accepted spellings of one element, which is the
 * aliasing this finding is about. Gating the input makes the builtins
 * canonical-in / canonical-out, so the gate is idempotent under composition.
 * Aborting form, because these are VALUE builtins — the split R-117 drew for
 * the EC value builtins and CL-BUG-095 set for the Point width.
 *
 * Everything below is EXECUTED on @bsv/sdk's `Spend`.
 */

const BB_P = 2013265921n;
const KB_P = 2130706433n;

function run(pushes: StackOp[], emitFn: (e: (op: StackOp) => void) => void) {
  const ops: StackOp[] = [...pushes];
  emitFn((op) => ops.push(op));
  // Leave a truthy top so `success` means "ran to completion", and read the
  // value through an explicit comparison in the callers that need one.
  const { scriptHex } = emitMethod({ name: 't', ops } as never) as { scriptHex: string };
  const r = new ScriptVM().executeHex(scriptHex);
  return { rejected: r.error !== undefined, stack: r.stack };
}

const num = (n: bigint) => ({ op: 'push', value: n }) as StackOp;

/** Decode a script number (little-endian, sign-magnitude) back to a bigint. */
function fromScriptNum(b: Uint8Array): bigint {
  if (b.length === 0) return 0n;
  const rev = Array.from(b).reverse();
  const neg = (rev[0]! & 0x80) !== 0;
  rev[0] = rev[0]! & 0x7f;
  let v = 0n;
  for (const byte of rev) v = (v << 8n) | BigInt(byte);
  return neg ? -v : v;
}

function result(pushes: StackOp[], emitFn: (e: (op: StackOp) => void) => void): bigint | null {
  const r = run(pushes, emitFn);
  if (r.rejected) return null;
  return fromScriptNum(r.stack[r.stack.length - 1]!);
}

const FIELDS = [
  {
    name: 'babybear', p: BB_P,
    add: emitBBFieldAdd, sub: emitBBFieldSub, mul: emitBBFieldMul, inv: emitBBFieldInv,
    ext4Mul: emitBBExt4Mul0, ext4Inv: emitBBExt4Inv0,
  },
  {
    name: 'koalabear', p: KB_P,
    add: emitKBFieldAdd, sub: emitKBFieldSub, mul: emitKBFieldMul, inv: emitKBFieldInv,
    ext4Mul: emitKBExt4Mul0, ext4Inv: emitKBExt4Inv0,
  },
] as const;

describe.each(FIELDS)('R-119 witness field elements are gated — $name', (F) => {
  const p = F.p;
  const pm1 = p - 1n;

  it('ASSERTION — an operand >= p is REJECTED on every scalar builtin', () => {
    for (const bad of [p, p + 5n, 2n * p]) {
      expect(run([num(bad), num(1n)], F.add).rejected, `add(${bad}, 1)`).toBe(true);
      expect(run([num(1n), num(bad)], F.add).rejected, `add(1, ${bad})`).toBe(true);
      expect(run([num(bad), num(1n)], F.sub).rejected).toBe(true);
      expect(run([num(1n), num(bad)], F.sub).rejected).toBe(true);
      expect(run([num(bad), num(1n)], F.mul).rejected).toBe(true);
      expect(run([num(1n), num(bad)], F.mul).rejected).toBe(true);
      expect(run([num(bad)], F.inv).rejected).toBe(true);
    }
  });

  it('ASSERTION — a NEGATIVE operand is REJECTED on every scalar builtin', () => {
    for (const bad of [-1n, -p, -(p + 1n)]) {
      expect(run([num(bad), num(0n)], F.add).rejected, `add(${bad}, 0)`).toBe(true);
      expect(run([num(0n), num(bad)], F.add).rejected).toBe(true);
      expect(run([num(bad), num(0n)], F.sub).rejected).toBe(true);
      expect(run([num(0n), num(bad)], F.sub).rejected).toBe(true);
      expect(run([num(bad), num(1n)], F.mul).rejected).toBe(true);
      expect(run([num(1n), num(bad)], F.mul).rejected).toBe(true);
      expect(run([num(bad)], F.inv).rejected).toBe(true);
    }
  });

  it('ASSERTION — each of the eight ext4Mul operands is gated independently', () => {
    const ok8 = [1n, 2n, 3n, 4n, 5n, 6n, 7n, pm1];
    expect(run(ok8.map(num), F.ext4Mul).rejected, 'canonical control').toBe(false);
    for (let i = 0; i < 8; i++) {
      for (const bad of [ok8[i]! + p, -1n, p]) {
        const v = [...ok8]; v[i] = bad;
        expect(run(v.map(num), F.ext4Mul).rejected, `operand ${i} = ${bad}`).toBe(true);
      }
    }
  });

  it('ASSERTION — each of the four ext4Inv operands is gated independently', () => {
    const ok4 = [1n, 2n, 3n, pm1];
    expect(run(ok4.map(num), F.ext4Inv).rejected, 'canonical control').toBe(false);
    for (let i = 0; i < 4; i++) {
      for (const bad of [ok4[i]! + p, -1n, p]) {
        const v = [...ok4]; v[i] = bad;
        expect(run(v.map(num), F.ext4Inv).rejected, `operand ${i} = ${bad}`).toBe(true);
      }
    }
  });

  // -------------------------------------------------------------------------
  // Controls. Each must STAY green; an over-strict gate reddens them, and each
  // accept is graded against a value computed here rather than by the compiler.
  // -------------------------------------------------------------------------

  it('CONTROL — every legal pair still computes the RIGHT value', () => {
    const mod = (x: bigint) => ((x % p) + p) % p;
    for (const [a, b] of [[0n, 0n], [0n, pm1], [pm1, 0n], [pm1, pm1], [5n, 7n], [1n, pm1]] as const) {
      expect(result([num(a), num(b)], F.add), `add(${a},${b})`).toBe(mod(a + b));
      expect(result([num(a), num(b)], F.sub), `sub(${a},${b})`).toBe(mod(a - b));
      expect(result([num(a), num(b)], F.mul), `mul(${a},${b})`).toBe(mod(a * b));
    }
  });

  it('CONTROL — inv still inverts, and inv(0) stays 0', () => {
    const powMod = (b: bigint, e: bigint) => {
      let r = 1n, base = b % p, exp = e;
      while (exp > 0n) { if (exp & 1n) r = (r * base) % p; base = (base * base) % p; exp >>= 1n; }
      return r;
    };
    for (const a of [1n, 3n, 12345n, pm1]) {
      expect(result([num(a)], F.inv), `inv(${a})`).toBe(powMod(a, p - 2n));
    }
    expect(result([num(0n)], F.inv)).toBe(0n);
  });

  it('BOUNDARY — p-1 is accepted and p is not, on both operands', () => {
    expect(run([num(pm1), num(pm1)], F.add).rejected).toBe(false);
    expect(run([num(p), num(0n)], F.add).rejected).toBe(true);
    expect(run([num(0n), num(p)], F.add).rejected).toBe(true);
    expect(run([num(pm1)], F.inv).rejected).toBe(false);
    expect(run([num(p)], F.inv).rejected).toBe(true);
  });

  it('CLOSURE — a gated builtin returns a value the next gate accepts', () => {
    // sub(add(a, b), b) === a in ONE script: the inner result is fed straight
    // into the outer builtin's gate. If the output ever escaped the field this
    // aborts, which is the property that makes gating the INPUT sufficient.
    for (const [a, b] of [[pm1, pm1], [0n, pm1], [5n, 7n]] as const) {
      const ops: StackOp[] = [num(a), num(b)];
      F.add((op) => ops.push(op));
      ops.push(num(b));
      F.sub((op) => ops.push(op));
      ops.push(num(a), { op: 'opcode', code: 'OP_NUMEQUAL' } as StackOp);
      const { scriptHex } = emitMethod({ name: 't', ops } as never) as { scriptHex: string };
      const r = new ScriptVM().executeHex(scriptHex);
      expect(r.error, `sub(add(${a},${b}),${b})`).toBeUndefined();
      expect(r.success).toBe(true);
    }
  });
});

/**
 * The reference interpreter must refuse on the SAME bound as the script — the
 * three-places rule `pow` (R-169) established, minus the folder (which does not
 * fold these builtins). Without it, the source-vs-script differential oracle
 * disagrees on every non-canonical witness and `TestContract` keeps telling an
 * author a spend works that the chain rejects.
 *
 * KoalaBear has no interpreter case at all — `kbFieldMul` falls through to
 * "Unknown function" — so only BabyBear is checked here.
 */
describe('R-119 the interpreter refuses what the script refuses', () => {
  const SRC = `
class BBDomain extends SmartContract {
  readonly expected: bigint;
  constructor(expected: bigint) { super(expected); this.expected = expected; }
  public verify(a: bigint, b: bigint) {
    assert(bbFieldMul(a, b) === this.expected);
  }
}
`;
  const c = () => TestContract.fromSource(SRC, { expected: 35n }, 'BBDomain.runar.ts');

  it('accepts a canonical pair and refuses >= p and negative', () => {
    const k = c();
    expect(k.call('verify', { a: 5n, b: 7n }).success).toBe(true);
    for (const bad of [BB_P, BB_P + 5n, -1n]) {
      const r = k.call('verify', { a: bad, b: 7n });
      expect(r.success, `a = ${bad}`).toBe(false);
      expect(r.error).toMatch(/not a BabyBear field element/);
    }
  });

  it('gates the second operand too', () => {
    const r = c().call('verify', { a: 5n, b: -1n });
    expect(r.success).toBe(false);
    expect(r.error).toMatch(/argument 1/);
  });
});
